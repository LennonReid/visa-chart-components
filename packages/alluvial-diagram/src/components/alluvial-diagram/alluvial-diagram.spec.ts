/**
 * Copyright (c) 2021, 2022 Visa, Inc.
 *
 * This source code is licensed under the MIT license
 * https://github.com/visa/visa-chart-components/blob/master/LICENSE
 *
 **/
import { newSpecPage, SpecPage } from '@stencil/core/testing';
import { AlluvialDiagram } from './alluvial-diagram';
import { AlluvialDiagramDefaultValues } from './alluvial-diagram-default-values';
import { scaleBand, scaleOrdinal, scaleLinear } from 'd3-scale';

// we need to bring in our nested components as well, was required to bring in the source vs dist folder to get it to mount
import { KeyboardInstructions } from '../../../node_modules/@visa/keyboard-instructions/src/components/keyboard-instructions/keyboard-instructions';
import { DataTable } from '../../../node_modules/@visa/visa-charts-data-table/src/components/data-table/data-table';

import Utils from '@visa/visa-charts-utils';
import UtilsDev from '@visa/visa-charts-utils-dev';

const { getColors, visaColors, formatStats, getContrastingStroke } = Utils;

const {
  asyncForEach,
  flushTransitions,
  unitTestAccessibility,
  unitTestEvent,
  unitTestGeneric,
  unitTestInteraction,
  unitTestTooltip
} = UtilsDev;

describe('<alluvial-diagram>', () => {
  // TECH DEBT: Need to revisit class-logic-testing post PURE function refactor.
  // Class-logic-testing is TDD and BDD friendly.
  describe('class-logic', () => {
    it('should build', () => {
      expect(new AlluvialDiagram()).toBeTruthy();
    });
  });

  describe('rendered-html', () => {
    let page: SpecPage;
    let component;

    // START:minimal props need to be passed to component
    const EXPECTEDLINKDATA = [
      { group: 'Remained', cat: 'Cat1 2018', target: 'Cat1 2019', value: 3010 },
      { group: 'Decreased', cat: 'Cat1 2018', target: 'Cat2 2019', value: 2754 },
      { group: 'Decreased', cat: 'Cat1 2018', target: 'Cat3 2019', value: 2812 },
      { group: 'Increased', cat: 'Cat2 2018', target: 'Cat1 2019', value: 909 },
      { group: 'Remained', cat: 'Cat2 2018', target: 'Cat2 2019', value: 12712 },
      { group: 'Decreased', cat: 'Cat2 2018', target: 'Cat3 2019', value: 3367 },
      { group: 'Increased', cat: 'Cat3 2018', target: 'Cat1 2019', value: 68 },
      { group: 'Increased', cat: 'Cat3 2018', target: 'Cat2 2019', value: 1133 },
      { group: 'Remained', cat: 'Cat3 2018', target: 'Cat3 2019', value: 6164 },
      { group: 'New', cat: 'CatNew 2018', target: 'Cat1 2019', value: 3148 },
      { group: 'New', cat: 'CatNew 2018', target: 'Cat2 2019', value: 7279 },
      { group: 'New', cat: 'CatNew 2018', target: 'Cat3 2019', value: 3684 }
    ];
    const EXPECTEDNODEDATA = [
      { did: 'Cat1 2018' },
      { did: 'Cat2 2018' },
      { did: 'Cat3 2018' },
      { did: 'CatNew 2018' },
      { did: 'Cat1 2019' },
      { did: 'Cat2 2019' },
      { did: 'Cat3 2019' }
    ];
    const EXPECTEDSOURCEACCESSOR = 'cat';
    const EXPECTEDTARGETACCESSOR = 'target';
    const EXPECTEDVALUEACCESSOR = 'value';
    const EXPECTEDNODEIDACCESSOR = 'did';
    const EXPECTEDGROUPACCESSOR = 'group';
    const MINVALUE = 68;
    const MAXVALUE = 12712;
    // END:minimal props need to be passed to component

    // disable accessibility validation to keep output stream(terminal) clean
    const EXPECTEDACCESSIBILITY = { ...AlluvialDiagramDefaultValues.accessibility, disableValidation: true };

    beforeEach(async () => {
      page = await newSpecPage({
        components: [AlluvialDiagram, KeyboardInstructions, DataTable],
        html: '<div></div>'
      });
      component = page.doc.createElement('alluvial-diagram');
      component.linkData = [...EXPECTEDLINKDATA];
      component.nodeData = [...EXPECTEDNODEDATA];
      component.sourceAccessor = EXPECTEDSOURCEACCESSOR;
      component.targetAccessor = EXPECTEDTARGETACCESSOR;
      component.valueAccessor = EXPECTEDVALUEACCESSOR;
      component.groupAccessor = EXPECTEDGROUPACCESSOR;
      component.nodeIDAccessor = EXPECTEDNODEIDACCESSOR;
      component.accessibility = EXPECTEDACCESSIBILITY;
      component.uniqueID = 'alluvial-diagram-unit-test';
      component.unitTest = true;
    });

    it('should build', () => {
      expect(new AlluvialDiagram()).toBeTruthy();
    });

    describe('render', () => {
      beforeEach(() => {
        // MOCK MATH.Random TO HANDLE UNIQUE ID CODE FROM ACCESSIBILITY UTIL
        jest.spyOn(global.Math, 'random').mockReturnValue(0.123456789);
      });

      afterEach(() => {
        // RESTORE GLOBAL FUNCTION FROM MOCK AFTER TEST
        jest.spyOn(global.Math, 'random').mockRestore();
      });

      it('should render with minimal props[data,accessors] given', async () => {
        // ACT
        page.root.appendChild(component);
        await page.waitForChanges();

        // flush labels for testing to ensure opacity of 1 on initial render
        const elements = page.doc.querySelectorAll('[data-testid=dataLabel]');
        await asyncForEach(elements, async element => {
          flushTransitions(element);
          await page.waitForChanges();
        });

        // ASSERT
        expect(page.root).toMatchSnapshot();
      });
    });

    // @test-notes: had to remove theme and data, due to restructuring of data for alluvial this test doesn't work as designed
    // for data we need to do d.data.data for alluvial vs just using __data__
    describe('generic test suite', () => {
      Object.keys(unitTestGeneric).forEach(test => {
        if (unitTestGeneric[test].prop !== 'data') {
          const innerTestProps = unitTestGeneric[test].testDefault
            ? { [unitTestGeneric[test].prop]: AlluvialDiagramDefaultValues[unitTestGeneric[test].prop] }
            : unitTestGeneric[test].prop === 'data'
            ? { data: EXPECTEDLINKDATA }
            : unitTestGeneric[test].testProps;
          const innerTestSelector =
            unitTestGeneric[test].testSelector === 'component-name'
              ? 'alluvial-diagram'
              : unitTestGeneric[test].testSelector === '[data-testid=mark]'
              ? '[data-testid=link]'
              : unitTestGeneric[test].testSelector;
          it(`${unitTestGeneric[test].prop}: ${unitTestGeneric[test].name}`, () =>
            unitTestGeneric[test].testFunc(component, page, innerTestProps, innerTestSelector));
        }
      });
      describe('data', () => {
        describe('uniqueId', () => {
          it('refer to generic results above for uniqueID tests', () => {
            expect(true).toBeTruthy();
          });
        });

        describe('data - generic data tests must be modified for alluvial-diagram', () => {
          const testSelector = '[data-testid=link]';
          it('data: custom data on load', async () => {
            // ACT
            page.root.appendChild(component);
            await page.waitForChanges();

            // ASSERT
            const elements = page.doc.querySelectorAll(testSelector);
            elements.forEach((element, i) => {
              expect(element['__data__'].data).toEqual(EXPECTEDLINKDATA[i]); // tslint:disable-line: no-string-literal
            });
          });
          it('data: custom data enter on update', async () => {
            // ARRANGE
            const BEGINNINGDATA = [EXPECTEDLINKDATA[0], EXPECTEDLINKDATA[1], EXPECTEDLINKDATA[2], EXPECTEDLINKDATA[3]];
            component.data = BEGINNINGDATA;

            // ACT RENDER
            page.root.appendChild(component);
            await page.waitForChanges();

            // ACT UPDATE
            component.data = EXPECTEDLINKDATA;
            await page.waitForChanges();

            // ASSERT
            const elements = page.doc.querySelectorAll(testSelector);
            elements.forEach((element, i) => {
              expect(element['__data__'].data).toEqual(EXPECTEDLINKDATA[i]); // tslint:disable-line: no-string-literal
            });
          });
          it('data: custom data exit on update', async () => {
            const EXITDATA = [EXPECTEDLINKDATA[1], EXPECTEDLINKDATA[0], EXPECTEDLINKDATA[2], EXPECTEDLINKDATA[3]];

            // ACT RENDER
            page.root.appendChild(component);
            await page.waitForChanges();

            // ACT UPDATE
            component.data = EXITDATA;
            await page.waitForChanges();

            // ASSERT
            const elements = page.doc.querySelectorAll(testSelector);
            elements.forEach((element, i) => {
              if (i + 1 <= EXITDATA.length) {
                expect(element['__data__'].data).toEqual(EXPECTEDLINKDATA[i]); // tslint:disable-line: no-string-literal
              } else {
                const lastTransitionKey = Object.keys(element['__transition'])[
                  Object.keys(element['__transition']).length - 1
                ]; // tslint:disable-line: no-string-literal
                const transitionName = element['__transition'][lastTransitionKey].name; // tslint:disable-line: no-string-literal
                expect(transitionName).toEqual('accessibilityAfterExit');
              }
            });
          });
        });
      });
    });
  });
});

// currently we have a bug where the test script will hang and not close properly
// if all tests pass, until this is resolved, we force fail a test at the end of the script
// to ensure tests close out and
describe('fake-test-fail-exit', () => {
  it('force fail to get tests to close jest testing script when all pass', () => {
    expect(false).toBeTruthy();
  });
});
