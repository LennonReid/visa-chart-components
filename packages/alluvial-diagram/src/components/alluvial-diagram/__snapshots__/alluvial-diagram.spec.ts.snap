// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<alluvial-diagram> rendered-html generic test suite accessibility: should not print accessibility console warnings and logs when accessibility.disableValidation set to true 1`] = `Array []`;

exports[`<alluvial-diagram> rendered-html generic test suite accessibility: should not print accessibility console warnings and logs when accessibility.disableValidation set to true 2`] = `Array []`;

exports[`<alluvial-diagram> rendered-html generic test suite accessibility: should not print accessibility console warnings and logs when accessibility.disableValidation set to true 3`] = `Array []`;

exports[`<alluvial-diagram> rendered-html generic test suite accessibility: should print accessibility console warnings and logs when accessibility.disableValidation defaults to false 1`] = `
Array [
  Array [
    "%c longDescription: Either accessibility.longDescription or accessibility.contextExplanation is required",
    "color: #C6003F;font-size:medium;",
  ],
  Array [
    "%c executiveSummary: Either accessibility.purpose or accessibility.executiveSummary is required",
    "color: #C6003F;font-size:medium;",
  ],
  Array [
    "%c elementsAreInterface: accessibility.elementsAreInterface needs a boolean value either true or false",
    "color: #C6003F;font-size:medium;",
  ],
]
`;

exports[`<alluvial-diagram> rendered-html generic test suite accessibility: should print accessibility console warnings and logs when accessibility.disableValidation defaults to false 2`] = `
Array [
  Array [
    "%c longDescription: Either accessibility.longDescription or accessibility.contextExplanation should have minimum 40 characters and a combined length between 40 and 500 characters",
    "font-size:medium;",
  ],
  Array [
    "%c executiveSummary: Either accessibility.purpose or accessibility.executiveSummary should have minimum 40 characters and a combined length between 40 and 250 characters",
    "font-size:medium;",
  ],
  Array [
    "%c statisticalNotes: accessibility.statisticalNotes should have length between 40 and 250 characters",
    "font-size:medium;",
  ],
  Array [
    "%c structureNotes: accessibility.structureNotes should have length between 40 and 250 characters",
    "font-size:medium;",
  ],
]
`;

exports[`<alluvial-diagram> rendered-html generic test suite accessibility: should print accessibility console warnings and logs when accessibility.disableValidation defaults to false 3`] = `
Array [
  Array [
    "%c alluvial-diagram-unit-test has accessibility warnings and other messages",
    "color: #C6003F;font-size:medium;",
  ],
]
`;

exports[`<alluvial-diagram> rendered-html render should render with minimal props[data,accessors] given 1`] = `
<alluvial-diagram id="alluvial-diagram-unit-test">
  <div>
    <div class="o-layout">
      <div class="vcl-accessibility-instructions" id="chart-instructions-alluvial-diagram-unit-test" style="position: absolute; width: 650px;">
        <p class="screen-reader-info vcl-region-label">
          Keyboard interactive alluvial-diagram, Titled: Alluvial Diagram Title. This section contains additional information about this chart. Pressing TAB will focus the keyboard instructions menu. Tabbing again takes you to the chart area.
        </p>
        <h2 class="screen-reader-info vcl-access-title">
          Chart title: Alluvial Diagram Title
        </h2>
        <h3 class="screen-reader-info vcl-access-subtitle">
          Chart subtitle: This is the alluvial diagram's subtitle
        </h3>
        <p class="screen-reader-info vcl-access-executive-summary" data-level="h3"></p>
        <p class="screen-reader-info vcl-access-purpose" data-level="h3"></p>
        <p class="screen-reader-info vcl-access-long-description" data-level="h3"></p>
        <p class="screen-reader-info vcl-access-context" data-level="h3"></p>
        <h3 class="screen-reader-info vcl-access-structure-heading">
          Structure
        </h3>
        <p class="screen-reader-info vcl-access-statistics" data-level="h4"></p>
        <h4 class="screen-reader-info vcl-access-chart-layout-heading">
          Chart Layout Description
        </h4>
        <p class="screen-reader-info vcl-access-layout">
          This is a alluvial-diagram with 12 links.
        </p>
        <p class="screen-reader-info vcl-access-xAxis"></p>
        <p class="screen-reader-info vcl-access-yAxis"></p>
        <p class="screen-reader-info vcl-access-notes" data-annotationlevel="h3" data-level="h4"></p>
      </div>
      <div class="o-layout--chart">
        <h2 aria-hidden="true" class="alluvial-main-title vcl-main-title" data-testid="main-title">
          Alluvial Diagram Title
        </h2>
        <p aria-hidden="true" class="alluvial-sub-title vcl-sub-title visa-ui-text--instructions" data-testid="sub-title">
          This is the alluvial diagram's subtitle
        </p>
        <keyboard-instructions>
          <div class="vcc-ki-keyboard-instructions-root" style="position: relative; width: 643.5px;">
            <div class="visa-viz-keyboard-instructions-outer-container" style="width: 416px;">
              <div class="vcc-ki-keyboard-instructions-header vcc-ki-minimize">
                <p class="vcc-ki-keyboard-heading">
                  Select to View Keyboard Instructions
                </p>
              </div>
              <div class="visa-viz-keyboard-instructions-button-wrapper">
                <button aria-expanded="false" aria-label="display keyboard instructions" class="vcc-ki-btn-icon vcc-ki-btn-icon--light-tiny visa-viz-keyboard-instructions-button" id="visa-viz-keyboard-instructions-button-alluvial-diagram-unit-test" role="button" tabindex="0" type="button">
                  <svg class="vcc-ki-icon--tiny" focusable="false" tabindex="-1">
                    <use href="#visa-viz-keyboard-instructions-info--tiny-alluvial-diagram-unit-test"></use>
                  </svg>
                </button>
                <svg class="vcc-ki-icons">
                  <symbol id="visa-viz-keyboard-instructions-info--tiny-alluvial-diagram-unit-test" viewbox="0 0 16 16">
                    <path d="M8.7,3.8a.68.68,0,0,1,.7.7.79.79,0,0,1-.8.8H8.5c-.4,0-.6-.3-.6-.7A.79.79,0,0,1,8.7,3.8ZM6.6,12.1l1.2-6H9l-1.2,6ZM14,8a6,6,0,1,0-6,6A6,6,0,0,0,14,8Zm2,0A8,8,0,1,1,8,0,8,8,0,0,1,16,8Z"></path>
                  </symbol>
                </svg>
              </div>
              <div class="visa-viz-keyboard-instructions-container" id="visa-viz-keyboard-instructions-container-alluvial-diagram-unit-test"></div>
            </div>
          </div>
        </keyboard-instructions>
        <div class="visa-viz-d3-alluvial-container" data-testid="chart-container" style="position: relative;">
          <div aria-label="Alluvial Diagram Title. This is the alluvial diagram's subtitle.  Navigate into the chart area by pressing ENTER. ID: alluvial-diagram-unit-test" class="VCL-controller" data-group="did" id="chart-area-alluvial-diagram-unit-test" role="application" tabindex="0" style="position: absolute; top: 0; left: 0; opacity: 0; pointer-events: none; width: 0px; height: 0px;">
            Interactive alluvial-diagram.
          </div>
          <svg class="vcl-svg" data-testid="root-svg" focusable="false" height="300" role="presentation" tabindex="-1" viewBox="0 0 650 300" width="650" style="overflow: hidden;">
            <g data-testid="margin-container" id="visa-viz-margin-container-g-alluvial-diagram-unit-test" transform="translate(6.5, 3)">
              <g data-testid="padding-container" id="visa-viz-padding-container-g-alluvial-diagram-unit-test" transform="translate(50, 20)">
                <g aria-hidden="true" class="alluvial-node-group" focusable="false" role="presentation">
                  <g class="node-wrapper" data-id="node-Cat1 2018" data-offset-group="true" data-testid="node" id="alluvial-diagram-unit-test-123457-123457-123457">
                    <rect class="alluvial-node" fill="#E4E4E4" height="29.899319727891154" opacity="0" stroke="#717171" stroke-width="1px" width="12" x="0" y="-2.842170943040401e-14"></rect>
                  </g>
                  <g class="node-wrapper" data-id="node-Cat2 2018" data-offset-group="true" data-testid="node" id="alluvial-diagram-unit-test-123457-123457-123457">
                    <rect class="alluvial-node" fill="#E4E4E4" height="59.22687074829936" opacity="0" stroke="#717171" stroke-width="1px" width="12" x="0" y="49.899319727891125"></rect>
                  </g>
                  <g class="node-wrapper" data-id="node-Cat3 2018" data-offset-group="true" data-testid="node" id="alluvial-diagram-unit-test-123457-123457-123457">
                    <rect class="alluvial-node" fill="#E4E4E4" height="25.67729591836735" opacity="0" stroke="#717171" stroke-width="1px" width="12" x="0" y="129.1261904761905"></rect>
                  </g>
                  <g class="node-wrapper" data-id="node-CatNew 2018" data-offset-group="true" data-testid="node" id="alluvial-diagram-unit-test-123457-123457-123457">
                    <rect class="alluvial-node" fill="#E4E4E4" height="49.19651360544216" opacity="0" stroke="#717171" stroke-width="1px" width="12" x="0" y="174.80348639455784"></rect>
                  </g>
                  <g class="node-wrapper" data-id="node-Cat1 2019" data-offset-group="true" data-testid="node" id="alluvial-diagram-unit-test-123457-123457-123457">
                    <rect class="alluvial-node" fill="#E4E4E4" height="24.875425170068013" opacity="0" stroke="#717171" stroke-width="1px" width="12" x="525" y="0"></rect>
                  </g>
                  <g class="node-wrapper" data-id="node-Cat2 2019" data-offset-group="true" data-testid="node" id="alluvial-diagram-unit-test-123457-123457-123457">
                    <rect class="alluvial-node" fill="#E4E4E4" height="83.24812925170062" opacity="0" stroke="#717171" stroke-width="1px" width="12" x="525" y="44.87542517006801"></rect>
                  </g>
                  <g class="node-wrapper" data-id="node-Cat3 2019" data-offset-group="true" data-testid="node" id="alluvial-diagram-unit-test-123457-123457-123457">
                    <rect class="alluvial-node" fill="#E4E4E4" height="55.876445578231255" opacity="0" stroke="#717171" stroke-width="1px" width="12" x="525" y="148.12355442176863"></rect>
                  </g>
                </g>
                <g aria-hidden="true" class="alluvial-link-group" focusable="false" role="presentation">
                  <path class="alluvial-link entering" cursor="default" d="M12,14.848554421768679C268.5,14.848554421768679,268.5,5.247023809523809,525,5.247023809523809" data-id="link-Cat1 2018-Cat1 2019" data-role="img" data-testid="link" fill="none" focusable="false" id="alluvial-diagram-unit-test-123457-123457-123457" role="presentation" stroke="#7c99b1" stroke-dasharray="-2 -2" stroke-dashoffset="-2" stroke-opacity="0.3" stroke-width="0.1" tabindex="-1"></path>
                  <path class="alluvial-link entering" cursor="default" d="M12,4.80076530612242C268.5,4.80076530612242,268.5,49.67619047619046,525,49.67619047619046" data-id="link-Cat1 2018-Cat2 2019" data-role="img" data-testid="link" fill="none" focusable="false" id="alluvial-diagram-unit-test-123457-123457-123457" role="presentation" stroke="#7c99b1" stroke-dasharray="-2 -2" stroke-dashoffset="-2" stroke-opacity="0.3" stroke-width="0.1" tabindex="-1"></path>
                  <path class="alluvial-link entering" cursor="default" d="M12,24.997448979591805C268.5,24.997448979591805,268.5,153.02542517006796,525,153.02542517006796" data-id="link-Cat1 2018-Cat3 2019" data-role="img" data-testid="link" fill="none" focusable="false" id="alluvial-diagram-unit-test-123457-123457-123457" role="presentation" stroke="#7c99b1" stroke-dasharray="-2 -2" stroke-dashoffset="-2" stroke-opacity="0.3" stroke-width="0.1" tabindex="-1"></path>
                  <path class="alluvial-link entering" cursor="default" d="M12,95.80293367346935C268.5,95.80293367346935,268.5,12.078613945578232,525,12.078613945578232" data-id="link-Cat2 2018-Cat1 2019" data-role="img" data-testid="link" fill="none" focusable="false" id="alluvial-diagram-unit-test-123457-123457-123457" role="presentation" stroke="#7c99b1" stroke-dasharray="-2 -2" stroke-dashoffset="-2" stroke-opacity="0.3" stroke-width="0.1" tabindex="-1"></path>
                  <path class="alluvial-link entering" cursor="default" d="M12,72.05884353741493C268.5,72.05884353741493,268.5,76.63647959183672,525,76.63647959183672" data-id="link-Cat2 2018-Cat2 2019" data-role="img" data-testid="link" fill="none" focusable="false" id="alluvial-diagram-unit-test-123457-123457-123457" role="presentation" stroke="#7c99b1" stroke-dasharray="-2 -2" stroke-dashoffset="-2" stroke-opacity="0.3" stroke-width="0.1" tabindex="-1"></path>
                  <path class="alluvial-link entering" cursor="default" d="M12,103.25684523809521C268.5,103.25684523809521,268.5,163.7966411564625,525,163.7966411564625" data-id="link-Cat2 2018-Cat3 2019" data-role="img" data-testid="link" fill="none" focusable="false" id="alluvial-diagram-unit-test-123457-123457-123457" role="presentation" stroke="#7c99b1" stroke-dasharray="-2 -2" stroke-dashoffset="-2" stroke-opacity="0.3" stroke-width="0.1" tabindex="-1"></path>
                  <path class="alluvial-link entering" cursor="default" d="M12,133.19481292517008C268.5,133.19481292517008,268.5,24.75688775510204,525,24.75688775510204" data-id="link-Cat3 2018-Cat1 2019" data-role="img" data-testid="link" fill="none" focusable="false" id="alluvial-diagram-unit-test-123457-123457-123457" role="presentation" stroke="#7c99b1" stroke-dasharray="-2 -2" stroke-dashoffset="-2" stroke-opacity="0.3" stroke-width="0.1" tabindex="-1"></path>
                  <path class="alluvial-link entering" cursor="default" d="M12,131.1012329931973C268.5,131.1012329931973,268.5,126.14851190476189,525,126.14851190476189" data-id="link-Cat3 2018-Cat2 2019" data-role="img" data-testid="link" fill="none" focusable="false" id="alluvial-diagram-unit-test-123457-123457-123457" role="presentation" stroke="#7c99b1" stroke-dasharray="-2 -2" stroke-dashoffset="-2" stroke-opacity="0.3" stroke-width="0.1" tabindex="-1"></path>
                  <path class="alluvial-link entering" cursor="default" d="M12,144.05841836734695C268.5,144.05841836734695,268.5,193.25493197278902,525,193.25493197278902" data-id="link-Cat3 2018-Cat3 2019" data-role="img" data-testid="link" fill="none" focusable="false" id="alluvial-diagram-unit-test-123457-123457-123457" role="presentation" stroke="#7c99b1" stroke-dasharray="-2 -2" stroke-dashoffset="-2" stroke-opacity="0.3" stroke-width="0.1" tabindex="-1"></path>
                  <path class="alluvial-link entering" cursor="default" d="M12,205.668537414966C268.5,205.668537414966,268.5,19.150765306122448,525,19.150765306122448" data-id="link-CatNew 2018-Cat1 2019" data-role="img" data-testid="link" fill="none" focusable="false" id="alluvial-diagram-unit-test-123457-123457-123457" role="presentation" stroke="#7c99b1" stroke-dasharray="-2 -2" stroke-dashoffset="-2" stroke-opacity="0.3" stroke-width="0.1" tabindex="-1"></path>
                  <path class="alluvial-link entering" cursor="default" d="M12,187.4922193877551C268.5,187.4922193877551,268.5,111.48473639455781,525,111.48473639455781" data-id="link-CatNew 2018-Cat2 2019" data-role="img" data-testid="link" fill="none" focusable="false" id="alluvial-diagram-unit-test-123457-123457-123457" role="presentation" stroke="#7c99b1" stroke-dasharray="-2 -2" stroke-dashoffset="-2" stroke-opacity="0.3" stroke-width="0.1" tabindex="-1"></path>
                  <path class="alluvial-link entering" cursor="default" d="M12,217.5780612244898C268.5,217.5780612244898,268.5,176.08792517006793,525,176.08792517006793" data-id="link-CatNew 2018-Cat3 2019" data-role="img" data-testid="link" fill="none" focusable="false" id="alluvial-diagram-unit-test-123457-123457-123457" role="presentation" stroke="#7c99b1" stroke-dasharray="-2 -2" stroke-dashoffset="-2" stroke-opacity="0.3" stroke-width="0.1" tabindex="-1"></path>
                </g>
                <g aria-hidden="true" class="alluvial-dataLabel-group" focusable="false" role="presentation">
                  <text class="alluvial-diagram-dataLabel" cursor="default" data-id="dataLabel-Cat1 2018" data-testid="dataLabel" data-translate-x="56.5" data-translate-y="23" data-use-dx="false" data-use-dy="false" data-x="12" data-y="14.949659863945548" dx="0.5em" dy="0.4em" filter="url(#VCL-t-s-f-ffffffalluvial-diagram-unit-test)" opacity="1" text-anchor="start" x="12" y="14.949659863945548">
                    8.6k
                  </text>
                  <text class="alluvial-diagram-dataLabel" cursor="default" data-id="dataLabel-Cat2 2018" data-testid="dataLabel" data-translate-x="56.5" data-translate-y="23" data-use-dx="false" data-use-dy="false" data-x="12" data-y="79.51275510204081" dx="0.5em" dy="0.4em" filter="url(#VCL-t-s-f-ffffffalluvial-diagram-unit-test)" opacity="1" text-anchor="start" x="12" y="79.51275510204081">
                    17k
                  </text>
                  <text class="alluvial-diagram-dataLabel" cursor="default" data-id="dataLabel-Cat3 2018" data-testid="dataLabel" data-translate-x="56.5" data-translate-y="23" data-use-dx="false" data-use-dy="false" data-x="12" data-y="141.96483843537416" dx="0.5em" dy="0.4em" filter="url(#VCL-t-s-f-ffffffalluvial-diagram-unit-test)" opacity="1" text-anchor="start" x="12" y="141.96483843537416">
                    7.4k
                  </text>
                  <text class="alluvial-diagram-dataLabel" cursor="default" data-id="dataLabel-CatNew 2018" data-testid="dataLabel" data-translate-x="56.5" data-translate-y="23" data-use-dx="false" data-use-dy="false" data-x="12" data-y="199.4017431972789" dx="0.5em" dy="0.4em" filter="url(#VCL-t-s-f-ffffffalluvial-diagram-unit-test)" opacity="1" text-anchor="start" x="12" y="199.4017431972789">
                    14.1k
                  </text>
                  <text class="alluvial-diagram-dataLabel" cursor="default" data-id="dataLabel-Cat1 2019" data-testid="dataLabel" data-translate-x="56.5" data-translate-y="23" data-use-dx="false" data-use-dy="false" data-x="525" data-y="12.437712585034006" dx="-0.5em" dy="0.4em" filter="url(#VCL-t-s-f-ffffffalluvial-diagram-unit-test)" opacity="1" text-anchor="end" x="525" y="12.437712585034006">
                    7.1k
                  </text>
                  <text class="alluvial-diagram-dataLabel" cursor="default" data-id="dataLabel-Cat2 2019" data-testid="dataLabel" data-translate-x="56.5" data-translate-y="23" data-use-dx="false" data-use-dy="false" data-x="525" data-y="86.49948979591832" dx="-0.5em" dy="0.4em" filter="url(#VCL-t-s-f-ffffffalluvial-diagram-unit-test)" opacity="1" text-anchor="end" x="525" y="86.49948979591832">
                    23.9k
                  </text>
                  <text class="alluvial-diagram-dataLabel" cursor="default" data-id="dataLabel-Cat3 2019" data-testid="dataLabel" data-translate-x="56.5" data-translate-y="23" data-use-dx="false" data-use-dy="false" data-x="525" data-y="176.06177721088426" dx="-0.5em" dy="0.4em" filter="url(#VCL-t-s-f-ffffffalluvial-diagram-unit-test)" opacity="1" text-anchor="end" x="525" y="176.06177721088426">
                    16k
                  </text>
                </g>
                <g aria-hidden="true" class="vcl-annotation-group" data-testid="annotation-group" focusable="false" role="presentation"></g>
                <g aria-hidden="true" class="editable vcl-annotation-group" data-testid="editable-annotation-group" focusable="false" role="presentation"></g>
              </g>
            </g>
            <defs>
              <filter class="vcl-inversion-filter" color-interpolation-filters="sRGB" id="inversion-alluvial-diagram-unit-test">
                <feColorMatrix in="SourceGraphic" type="matrix" values="-1 0 0 0 1 
        0 -1 0 0 1 
        0 0 -1 0 1
        0 0 0 1 0"></feColorMatrix>
                <feColorMatrix type="hueRotate" values="180"></feColorMatrix>
              </filter>
              <filter class="VCL-t-s-f-ffffff" id="VCL-t-s-f-ffffffalluvial-diagram-unit-test">
                <feMorphology in="SourceAlpha" operator="dilate" radius="1.5" result="dilatedText"></feMorphology>
                <feFlood flood-color="#ffffff" flood-opacity="1" result="whiteTextFlood"></feFlood>
                <feComposite in="whiteTextFlood" in2="dilatedText" operator="in" result="textOutline"></feComposite>
                <feMerge>
                  <feMergeNode in="textOutline"></feMergeNode>
                  <feMergeNode in="textOutline"></feMergeNode>
                  <feMergeNode in="textOutline"></feMergeNode>
                  <feMergeNode in="SourceGraphic"></feMergeNode>
                </feMerge>
              </filter>
            </defs>
          </svg>
        </div>
        <div class="alluvial-tooltip vcl-tooltip" data-testid="tooltip-container" style="display: block; position: absolute; padding: 0.5rem; opacity: 0; background-color: #ffffff; font-size: 14px; font-weight: 400; color: #222222; border: 1px solid #565656; border-radius: 3px; pointer-events: none; min-width: 80px; max-width: 300px;">
          <p style="margin: 0;"></p>
        </div>
        <data-table>
          <div class="visa-viz-data-table-outer-container" data-testid="data-table-outer-container">
            <div class="visa-viz-button-wrapper" style="padding-bottom: 0px;">
              <button aria-expanded="false" aria-label="display data table" class="vcc-btn-icon vcc-btn-icon--light-tiny visa-viz-data-table-button" id="visa-viz-data-table-button-alluvial-diagram-unit-test" role="button" tabindex="0" type="button" style="transform: translate(21.5px,
                -40px);">
                <svg class="vcc-icon--tiny" focusable="false" tabindex="-1">
                  <use href="#visa-viz-view-grid--tiny-alluvial-diagram-unit-test"></use>
                </svg>
              </button>
              <svg class="vcc-icons">
                <symbol id="visa-viz-view-grid--tiny-alluvial-diagram-unit-test" viewbox="0 0 16 16">
                  <path d="M5,5 L5,2 L2,2 L2,5 L5,5 Z M5,7 L2,7 C0.894905844,7 0,6.1044757 0,5 L0,2 C0,0.8955243 0.894905844,3.55271368e-15 2,3.55271368e-15 L5,3.55271368e-15 C6.10509416,3.55271368e-15 7,0.8955243 7,2 L7,5 C7,6.1044757 6.10509416,7 5,7 Z M14,5 L14,2 L11,2 L11,5 L14,5 Z M14,7 L11,7 C9.89490584,7 9,6.1044757 9,5 L9,2 C9,0.8955243 9.89490584,3.55271368e-15 11,3.55271368e-15 L14,3.55271368e-15 C15.1050942,3.55271368e-15 16,0.8955243 16,2 L16,5 C16,6.1044757 15.1050942,7 14,7 Z M5,14 L5,11 L2,11 L2,14 L5,14 Z M5,16 L2,16 C0.894905844,16 0,15.1044757 0,14 L0,11 C0,9.8955243 0.894905844,9 2,9 L5,9 C6.10509416,9 7,9.8955243 7,11 L7,14 C7,15.1044757 6.10509416,16 5,16 Z M14,14 L14,11 L11,11 L11,14 L14,14 Z M14,16 L11,16 C9.89490584,16 9,15.1044757 9,14 L9,11 C9,9.8955243 9.89490584,9 11,9 L14,9 C15.1050942,9 16,9.8955243 16,11 L16,14 C16,15.1044757 15.1050942,16 14,16 Z"></path>
                  <g fill="none" fill-rule="evenodd" stroke="none" stroke-width="1">
                    <mask fill="#fff">
                      <use xlink:href="#path-1"></use>
                    </mask>
                    <use fill="#000" fill-rule="nonzero" xlink:href="#path-1"></use>
                  </g>
                </symbol>
              </svg>
            </div>
            <div class="visa-viz-data-table-container" data-testid="data-table-container" id="visa-viz-data-table-container-alluvial-diagram-unit-test"></div>
          </div>
        </data-table>
      </div>
    </div>
  </div>
</alluvial-diagram>
`;
